lib/ImageWorker.js

```diff
@@ -34,33 +29,72 @@ const wrappedComponent = WrappedComponent => props => {
   return <WrappedComponent {...props} />;
 };
 
+/** Have we initiated the worker pool already? */
+const workerPoolCreated = false;
+const workerPool = [];
+
+function createWorkerPool() {
+  const blobURL = URL.createObjectURL(
+    new Blob([webWorkerScript], { type: 'application/javascript' })
+  );
+  for (let i = 0; i < window.navigator.hardwareConcurrency || 4; ++i) {
+    workerPool.push({
+      worker: blobURL,
+      inUse: false,
+      i
+    })
+  }
+}
+
+/** Returns the next available worker. */
+function getNextWorker() {
+  for (let i = 0; i < workerPool.length; ++i) {
+    const worker = workerPool[i];
+    if (!worker.inUse) {
+      worker.inUse = true;
+      return worker
+    }
+  }
+  // no free found, so we just return the first one
+  return workerPool[0];
+}
+
+/** Marks worker `index` as available. */
+function setFree(index: int) {
+  workerPool[index].inUse = false
+}
+
+
 class ImageWorker extends Component<ImageWorkerProps, ImageWorkerState> {
   image: HTMLImageElement;
-  worker = new Worker(URL.createObjectURL(
-    new Blob([webWorkerScript], { type: 'application/javascript' })
-  ))
 
   state = {
     isLoading: true,
     imgSrc: ''
   }
   constructor(props: ImageWorkerProps) {
     super(props);
-    this.worker.onmessage = (event: Object) => {
```

Comment:
What was the purpose of removing the listener from the constructor and moving it to CDM?

Reply:
I moved everything to the constructor back again. My original idea was that it should only load the image when truly needed, but the constructor is called before CDM.